R/type_factories.R

```diff
@@ -0,0 +1,155 @@
+assert_dependency <- function(library_name) {
+  flag <- requireNamespace(library_name, quietly = TRUE)
+  if (isFALSE(flag)) {
+    stop(sprintf("Please install the `%s` package.", library_name), .call = FALSE)
+  }
+  return(invisible(TRUE))
+}
+
+
+#' Type factory: Spline
+#'
+#' Call this function in the `type` argument of `tinyplot()` to display a 
+#' spline-smoothed version of the relationship between `y` and `x`.
+#'
+#' @inheritParams tinyplot
+#' @inheritParams mgcv::s
+#' @inheritParams marginaleffects::predictions
+#' @param conf_level FALSE or a numeric value between 0 and 1 determining the size of confidence intervals to display as ribbons.
+#' @details
+#' First, we fit a model using the [mgcv::gam] and [mgcv::s] function. The predicted values of `y` are then obtained by calling [marginaleffects::predictions].
+#'
+#' @examples
+#' plt(Sepal.Length ~ Petal.Length, type = type_spline(), data = iris)
+#' @export
+type_spline = function(x, y, conf_level = 0.95, k = -1, fx = FALSE, bs = "tp", ...) {
+    se <- if (isFALSE(conf_level)) FALSE else TRUE
+    conf_level <- if (isFALSE(conf_level)) 0.95 else conf_level
+    assert_dependency("mgcv")
+    assert_dependency("marginaleffects")
+    fun = function(x, y, ...) {
+        if (missing(x) || missing(y)) {
+            if (isTRUE(se)) {
+                return("ribbon")
+            } else {
+                return("l")
+            }
+        }
+        dat = data.frame(x, y)
+        fit = mgcv::gam(y ~ s(x, k = k, fx = fx, bs = bs), data = dat)
+        nd = data.frame(x = seq(min(x), max(x), length.out = 1000))
+        p = marginaleffects::predictions(fit, vcov = se, conf_level = conf_level, newdata = nd)
```

Comment:
It's obviously more work, but we should be able to stick to `stats::predict` and avoid the extra dependency with something like:

```r
p = predict(fit, se.fit = TRUE)
rngx = range(fit$data[["x"]], na.rm = TRUE)
xpred = seq(rngx[1], rngx[2], length.out = 100)
nd = data.frame(x = xpred)
p = as.data.frame(predict(fit, newdata = nd, se.fit = TRUE))
crit = qt(1 - (1-0.95)/2, fit$df.residual)
p[["lwr"]] = p[["fit"]] - (crit * p[["se.fit"]])
p[["upr"]] = p[["fit"]] + (crit * p[["se.fit"]])
p[["x"]] = xpred
```

Reply:
What about mgcv? If we don't Suggest that to {splines}, we lose cool functionality.

Reply:
I'm fine with `mgcv` since it's part of the Recommended channel.

(To clarify, though: you're saying passing `splines` to, say, `lm` as an alternative?)

Reply:
Sorry, I mistyped. I meant you probably could achieve the same thing with the `splines` package, but either `splines` of `mgcv` is necessary for this functionality. IME it works better than `loess()`, so it's nice to have this here. Recommended seems like a clear demarcation.